File: alpamon/io/queue.py
import time
from queue import PriorityQueue


RETRY_LIMIT = 5


class PriorityEntry:
    def __init__(self, priority, method, url, data, due=None, expiry=None):
        self.priority = priority
        self.method = method
        self.url = url
        self.data = data
        self.due = time.time() if due is None else due  # entries will not be handled until due
        self.expiry = expiry      # expired entries are ignored
        self.retry = RETRY_LIMIT  # remaning retry count

    def __lt__(self, other):
        if self.priority == other.priority:
            return self.due < other.due
        else:
            return self.priority < other.priority

    def __str__(self):
        return '(%d, %s, %s)' % (self.priority, self.due, self.url)


class RequestQueue:
    def __init__(self):
        self.queue = PriorityQueue(maxsize=10*60*60)  # 10 entries/second * 1h
    
    def request(self, method, url, json=None, priority=10, due=None, expiry=None, **kwargs):
        self.queue.put(PriorityEntry(priority, method, url, json, due, expiry))

    def post(self, url, json=None, priority=10, **kwargs):
        return self.request('POST', url, json=json, priority=priority, **kwargs)

    def patch(self, url, json=None, priority=10, **kwargs):
        return self.request('PATCH', url, json=json, priority=priority, **kwargs)

    def put(self, url, json=None, priority=10, **kwargs):
        return self.request('PUT', url, json=json, priority=priority, **kwargs)


rqueue = RequestQueue()

File: pkg/scheduler/queue.go
package scheduler

import (
	"github.com/adrianbrad/queue"
	"github.com/rs/zerolog/log"
	"net/http"
	"sync"
	"time"
)

var (
	Rqueue *RequestQueue
)

const (
	RetryLimit   = 5
	MaxQueueSize = 10 * 60 * 60 // 10 entries/second * 1h
)

func newRequestQueue() {
	Rqueue = &RequestQueue{
		queue: queue.NewPriority(
			[]PriorityEntry{},
			lessFunc,
			queue.WithCapacity(MaxQueueSize),
		),
		cond: sync.NewCond(&sync.Mutex{}),
	}
}

// less function of PriorityQueue
func lessFunc(elem, otherElem PriorityEntry) bool {
	if elem.priority == otherElem.priority {
		return elem.due.Before(otherElem.due) // elem.Due < otherElem.Due
	}
	return elem.priority < otherElem.priority
}

func (rq *RequestQueue) request(method, url string, data interface{}, priority int, due time.Time) {
	// time.Time{} : 0001-01-01 00:00:00 +0000 UTC
	if due.IsZero() {
		due = time.Now()
	}

	entry := PriorityEntry{
		priority: priority,
		method:   method,
		url:      url,
		data:     data,
		due:      due,
		// expiry:
		retry: RetryLimit,
	}

	err := rq.queue.Offer(entry)
	if err != nil {
		log.Error().Err(err).Msg("Error offering priority entry")
	}

	rq.cond.Signal()
}

func (rq *RequestQueue) Post(url string, data interface{}, priority int, due time.Time) {
	rq.request(http.MethodPost, url, data, priority, due)
}

func (rq *RequestQueue) Patch(url string, data interface{}, priority int, due time.Time) {
	rq.request(http.MethodPatch, url, data, priority, due)
}

func (rq *RequestQueue) Put(url string, data interface{}, priority int, due time.Time) {
	rq.request(http.MethodPut, url, data, priority, due)
}

func (rq *RequestQueue) Delete(url string, data interface{}, priority int, due time.Time) {
	rq.request(http.MethodDelete, url, data, priority, due)
}
